subsys/lorawan/lorawan.c

```diff
@@ -0,0 +1,483 @@
+		return "Busy ping-slot window time";
+	case LORAMAC_STATUS_BUSY_UPLINK_COLLISION:
+		return "Busy uplink collision";
+	case LORAMAC_STATUS_CRYPTO_ERROR:
+		return "Crypto error";
+	case LORAMAC_STATUS_FCNT_HANDLER_ERROR:
+		return "FCnt handler error";
+	case LORAMAC_STATUS_MAC_COMMAD_ERROR:
+		return "MAC command error";
+	case LORAMAC_STATUS_CLASS_B_ERROR:
+		return "ClassB error";
+	case LORAMAC_STATUS_CONFIRM_QUEUE_ERROR:
+		return "Confirm queue error";
+	case LORAMAC_STATUS_MC_GROUP_UNDEFINED:
+		return "Multicast group undefined";
+	case LORAMAC_STATUS_ERROR:
+		return "Unknown error";
+	default:
+		return NULL;
+	}
+}
+
+const char *eventinfo2str(int status)
+{
+	switch (status) {
+	case LORAMAC_EVENT_INFO_STATUS_OK:
+		return "OK";
+	case LORAMAC_EVENT_INFO_STATUS_ERROR:
+		return "Error";
+	case LORAMAC_EVENT_INFO_STATUS_TX_TIMEOUT:
+		return "Tx timeout";
+	case LORAMAC_EVENT_INFO_STATUS_RX1_TIMEOUT:
+		return "Rx 1 timeout";
+	case LORAMAC_EVENT_INFO_STATUS_RX2_TIMEOUT:
+		return "Rx 2 timeout";
+	case LORAMAC_EVENT_INFO_STATUS_RX1_ERROR:
+		return "Rx1 error";
+	case LORAMAC_EVENT_INFO_STATUS_RX2_ERROR:
+		return "Rx2 error";
+	case LORAMAC_EVENT_INFO_STATUS_JOIN_FAIL:
+		return "Join failed";
+	case LORAMAC_EVENT_INFO_STATUS_DOWNLINK_REPEATED:
+		return "Downlink repeated";
+	case LORAMAC_EVENT_INFO_STATUS_TX_DR_PAYLOAD_SIZE_ERROR:
+		return "Tx DR payload size error";
+	case LORAMAC_EVENT_INFO_STATUS_DOWNLINK_TOO_MANY_FRAMES_LOSS:
+		return "Downlink too many frames loss";
+	case LORAMAC_EVENT_INFO_STATUS_ADDRESS_FAIL:
+		return "Address fail";
+	case LORAMAC_EVENT_INFO_STATUS_MIC_FAIL:
+		return "MIC fail";
+	case LORAMAC_EVENT_INFO_STATUS_MULTICAST_FAIL:
+		return "Multicast fail";
+	case LORAMAC_EVENT_INFO_STATUS_BEACON_LOCKED:
+		return "Beacon locked";
+	case LORAMAC_EVENT_INFO_STATUS_BEACON_LOST:
+		return "Beacon lost";
+	case LORAMAC_EVENT_INFO_STATUS_BEACON_NOT_FOUND:
+		return "Beacon not found";
+	default:
+		return NULL;
+	}
+}
+
+/*
+ * MAC status and Event status to Zephyr error code conversion.
+ * Direct mapping is not possible as statuses often indicate the domain from
+ * which the error originated rather than its cause or meaning. -EINVAL has been
+ * used as a general error code because those usually result from incorrect
+ * configuration.
+ */
+const int mac_status_to_errno[] = {
+	0,			/* LORAMAC_STATUS_OK */
+	-EBUSY,			/* LORAMAC_STATUS_BUSY */
+	-ENOPROTOOPT,		/* LORAMAC_STATUS_SERVICE_UNKNOWN */
+	-EINVAL,		/* LORAMAC_STATUS_PARAMETER_INVALID */
+	-EINVAL,		/* LORAMAC_STATUS_FREQUENCY_INVALID */
+	-EINVAL,		/* LORAMAC_STATUS_DATARATE_INVALID */
+	-EINVAL,		/* LORAMAC_STATUS_FREQ_AND_DR_INVALID */
+	-ENOTCONN,		/* LORAMAC_STATUS_NO_NETWORK_JOINED */
+	-EMSGSIZE,		/* LORAMAC_STATUS_LENGTH_ERROR */
+	-EPFNOSUPPORT,		/* LORAMAC_STATUS_REGION_NOT_SUPPORTED */
+	-EMSGSIZE,		/* LORAMAC_STATUS_SKIPPED_APP_DATA */
+	-ECONNREFUSED,		/* LORAMAC_STATUS_DUTYCYCLE_RESTRICTED */
```

Comment:
-EBUSY (or posisbly -EAGAIN) might be more appropriate here since this happens when the PHY throttles the application to make sure the duty cycle doesn't exceed what's allowed for the channel. The expectation should be that the application will be able to re-transmit in the future. ECONNREFUSED seems more appropriate for cases where there is an explicit NAK from the server.

Another option would be for send to block until the channel becomes available, but that's probably best controlled by a flag.